### PR TITLE
Feature/jmh+improvements

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR provides the following:

* bug fix for the case when the assembly hooks are enabled in Reactor so the cancellation may end up exceptionally with NPE
* provides some fixes in tests
* improves the performance of `AbstractStreamObserverAndPublisher` by specifying the lowerTild for next prefetch (it means that we don't have to wait for full prefetch fulfillment so upstream have some space to send, and downstream does not call request to frequently)
* provides reference JMH benchmarks
* updates versions of libs

### CODE REVIEW

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java [hunk 1]
```diff
@@ -126,6 +127,7 @@ public void request(long n) {
             Consumer<CallStreamObserver<?>> onSubscribe,
             Runnable onTerminate) {
         this.prefetch = prefetch;
+        this.limit = (prefetch * 2) / 3;
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java]: Why multiply by 2/3? Maybe the factor should be a well named constant.

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java [hunk 2]
```diff
@@ -67,6 +67,7 @@ public void request(long n) {
 
     private final Queue<T> queue;
     private final int prefetch;
+    private final int limit;
```
Comment (rmichela) [inline common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java]: Can you add a comment explaining the difference between `prefetch` and `limit`?

File: reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkGRpcServerServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            }
+
+            @Override
+            public void onCompleted() {
+
+            }
+        };
+    }
+
+    //    @Override
```
Comment (rmichela) [inline reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkGRpcServerServiceImpl.java]: Is the commented out code needed?

File: reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkReactorServerServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        return request.then(responseMono);
+    }
+
+    @Override
+    public Flux<Messages.SimpleResponse> streamingBothWays(Flux<Messages.SimpleRequest> request) {
+        request.subscribe();
+        return responseFlux;
+    }
+
+    //    @Override
```
Comment (rmichela) [inline reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkReactorServerServiceImpl.java]: Is the commented out code needed?

**Comment by rmichela** [comment]
Can you please put bullet points in the PR description for the fixes you made and the dependency upgrades you made?

**Comment by rmichela** [comment]
Failed tests:  
```
clientToServerBackpressure(com.salesforce.rxgrpc.BackpressureIntegrationTest): expected:<[1]L> but was:<[0]L>
serverToClientBackpressure(com.salesforce.rxgrpc.BackpressureIntegrationTest): expected:<[1]L> but was:<[2]L>
```

File: reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkReactorServerServiceImpl.java (hunk 4 above)
Comment (OlegDokuka) [reply reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkReactorServerServiceImpl.java]: cleaned

File: reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkGRpcServerServiceImpl.java (hunk 3 above)
Comment (OlegDokuka) [reply reactor/reactor-grpc-jmh/src/test/java/com/salesforce/reactorgrpc/jmh/BenchmarkGRpcServerServiceImpl.java]: cleaned

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java (hunk 1 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java]: Fixed. That factor specifies the number of elements after which the next prefetch should happen. It is better to have limit lower than prefetch (which is the number of elements requested at first) so the upstream will have some number of elements to send before prefetch is fulfilled by the upstream. It allows avoiding the case when the sender is blocked waiting for the next request.

File: common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java (hunk 2 above)
Comment (OlegDokuka) [reply common/reactive-grpc-common/src/main/java/com/salesforce/reactivegrpc/common/AbstractStreamObserverAndPublisher.java]: added

**Comment by OlegDokuka** [comment]
@rmichela fixed all comments

**Pull request merged**